### PR TITLE
Version 1

### PR DESCRIPTION
**Pull request by AbsentSemicolon**
This brings in work for getting the repository to be usable.

Please Squash the request when accepting.

### CODE REVIEW

File: .eslint.yaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+    es6: true
+    jasmine: true
+    node: true
+extends: eslint:recommended
+rules:
+    arrow-body-style:
+        - error
+        - always
+    arrow-parens: error
+    arrow-spacing: error
```
Comment (fidian) [inline .eslint.yaml]: Why isn't this using the same eslint.yaml as opentoken?

File: .eslintignore [hunk 2]
```diff
@@ -0,0 +1,3 @@
+coverage/
+node_modules/
+node-test-helpers/
```
Comment (fidian) [inline .eslintignore]: Why is node-test-helpers being excluded from linting?

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
+Contributing
+============
+
+A guide to contributing to Node Test Helpers.
+
+## Preparing to Fork
```
Comment (fidian) [inline CONTRIBUTING.md]: I believe our standard is to use hyphens for level 2 headings.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+The first thing you'll want to do is making sure your IDE/editor of choice has a few things:
+1. [ESLint] is installed and you can utilize the [ESLint] configuration created in this repo. If you are unfamiliar with [ESLint], check out the site and see how great it is at keeping code consistent and check for some bad lines of code.
```
Comment (fidian) [inline CONTRIBUTING.md]: Add a newline between lists and paragraphs.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+2. [EditorConfig] is installed. Again, if unfamiliar with this tool, please take a few moments to familiarize yourself with it as this is also a handy tool.
+
+## Forking the Repository
+1. Fork the repository from https://github/tests-always-included/jasmine-node-helpers. Please do not fork from a fork.
```
Comment (fidian) [inline CONTRIBUTING.md]: Add a newline between lists and headings.

Really, add a newline to separate lists from anything that isn't part of the list.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,23 @@
+Contributing
+============
+
+A guide to contributing to Node Test Helpers.
+
+## Preparing to Fork
+The first thing you'll want to do is making sure your IDE/editor of choice has a few things:
```
Comment (fidian) [inline CONTRIBUTING.md:8]: Add a newline to separate headers from paragraph text.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+2. [EditorConfig] is installed. Again, if unfamiliar with this tool, please take a few moments to familiarize yourself with it as this is also a handy tool.
+
+## Forking the Repository
+1. Fork the repository from https://github/tests-always-included/jasmine-node-helpers. Please do not fork from a fork.
+2. Create a branch off of `master`. You don't want to work on `master` as this can cause issues later on.
+3. Write the code as you normally would, but make sure to adhere to the guidelines set out in the [ESLint] and [EditorConfig] guides.
+    * You should also include JSDoc blocks where applicable. [ESLint] will warn you where they should be and what is missing from them.
+4. Write tests. We want to maintain a 100% code coverage, so your pull request probably won't get approved if something isn't tested. This includes `else` branches of `if` statements as well.
+5. Submit a pull request at https://github.com/tests-always-included/jasmine-node-helpers/pulls.
```
Comment (fidian) [inline CONTRIBUTING.md]: Markdown formatters may not make this into a link.  I think this should be link-ified.

File: README.md [hunk 8]
```diff
@@ -1 +1,116 @@
-# node-test-helpers
\ No newline at end of file
+Jasmine-Node Helpers
+====================
+*Helping you test your [Node.js] applications, one helper at a time.*
```
Comment (fidian) [inline README.md:4]: Newline above.

File: README.md [hunk 9]
```diff
@@ -1 +1,116 @@
...
+
+[![Build Status][travis-image]][Travis CI]
+[![Dependencies][dependencies-image]][Dependencies]
+[![Dev Dependencies][devdependencies-image]][Dev Dependencies]
+[![codecov.io][codecov-image]][Code Coverage]
+
+About
+-----
+This contains a series of [Jasmine]1.3 helpers when using [Jasmine-Node] to facilitate testing certain functionality more easily. These help to detect when a test fails but should have passed, turn middleware into a promise and makes it easier to use promises in tests. More explanation is below for each helper.
```
Comment (fidian) [inline README.md]: Newline above.  Jasmine also needs a space after it and before "1.3".

File: README.md [hunk 10]
```diff
@@ -1 +1,116 @@
...
+
+*Note: If you are not using [Jasmine-Node] these helpers aren't going to help your testing suite very much.*
+
+How to Use
+----------
+Include the package in your `package.json` file.
```
Comment (fidian) [inline README.md:21]: Newline above.

File: README.md [hunk 11]
```diff
@@ -1 +1,116 @@
...
+
+```
+npm install --save-dev node-test-helpers
```
Comment (fidian) [inline README.md]: Indentation instead of code fences, per our day-long discussion about code.

File: README.md [hunk 12]
```diff
@@ -1 +1,116 @@
...
+```
+
+Then you need to be able to include the helper in your testing directory or where you'd like to run tests.
+
+You can do this by creating a symbolic link to files contained. You'll want to link to the files and not the folder. [Jasmine] doesn't seem to pick up on the files if only the folder is linked. Below is an example of setting up a symbolic link. Make sure to include the folder in your `.gitignore` file so you don't commit the files as this folder will show up in your status.
+
+```
+// Linux/MacOS
```
Comment (fidian) [inline README.md]: Indentation instead of fences.

File: README.md [hunk 13]
```diff
@@ -1 +1,116 @@
...
+mkdir ./path/to/project/test-folder/jasmine-node-helpers
+
+ln -s /path/to/project/from/root/node_modules/jasmine-node-helpers/lib/* ./path/to/project/test-folder/jasmine-node-helpers/
```
Comment (fidian) [inline README.md]: This is a very long line.  I suggest you `cd /path/to/project` and try to shrink the line a bit.

File: README.md [hunk 14]
```diff
@@ -1 +1,116 @@
...
+
+// Windows
+// If you can figure out how to make symlink work, good on you, make a [fork](CONTRIBUTING.md) and update those instructions please.
+```
+
+jasmine.fail
```
Comment (fidian) [inline README.md]: I'd prefer that this header looked like this:

```
`jasmine.fail([actual], [expected])`
```

File: README.md [hunk 15]
```diff
@@ -1 +1,116 @@
...
+------------
+This is most useful when testing a promise which you are wanting to check the test is getting to the reject successfully, but want to fail in case code changes which makes the promise resolve instead of reject. This will make the test fail so it can be caught and code can be diagnosed.
```
Comment (fidian) [inline README.md]: Newline above.

Mention that you can pass parameters and they are only used to generate error messages that make sense for the test.  Typically they don't need to get used.

File: README.md [hunk 16]
```diff
@@ -1 +1,116 @@
...
+
+```js
+describe("a suite", () => {
```
Comment (fidian) [inline README.md]: Indentation instead of fences.

File: README.md [hunk 17]
```diff
@@ -1 +1,116 @@
...
+    it("a test", () => {
+        return someClass.someMethodAsync().then(jasmine.fail, () => {
+            expect(someVar).toBe("this");
+        });
+    });
+});
+```
+
+jasmine.middlewareToPromise
```
Comment (fidian) [inline README.md]: Again, I prefer to show the arguments here.

```
`jasmine.middlewareToPromise(middlewareFunction)`
```

File: README.md [hunk 18]
```diff
@@ -1 +1,116 @@
...
+---------------------------
+Used when wanting to make a piece of middleware used in setting up a node server and you want to make it to be run synchronously with your tests to get a response back straight aways without waiting for a real promise to resolve.
```
Comment (fidian) [inline README.md]: Newline above.

File: README.md [hunk 19]
```diff
@@ -1 +1,116 @@
...
+
+This a bit of a complicated process and your middleware will need to be written in the style as the example and test is in, but should make it easy to test your middleware.
```
Comment (fidian) [inline README.md]: I don't understand the second sentence.  I believe you are trying to say that your middleware will still need to be written in the traditional style with a `done` or `next` callback at the end, but at least testing can be made much easier and will work hand-in-hand with the promises helper.

File: README.md [hunk 20]
```diff
@@ -1 +1,116 @@
...
+
+```js
+describe("a file", () => {
```
Comment (fidian) [inline README.md]: Indentation, not fences.

File: README.md [hunk 21]
```diff
@@ -1 +1,116 @@
...
+// If you can figure out how to make symlink work, good on you, make a [fork](CONTRIBUTING.md) and update those instructions please.
+```
+
+jasmine.fail
+------------
+This is most useful when testing a promise which you are wanting to check the test is getting to the reject successfully, but want to fail in case code changes which makes the promise resolve instead of reject. This will make the test fail so it can be caught and code can be diagnosed.
+
+```js
+describe("a suite", () => {
+    it("a test", () => {
```
Comment (fidian) [inline README.md]: Would you change "a test" to some action that you're pretending to test?  `it("rejects the returned promise", () => {`

File: README.md [hunk 22]
```diff
@@ -1 +1,116 @@
...
+    describe("a suite", () => {
+        it("a test", () => {
+            var middlewareAsync;
+
+            middlewareAsync = jasmine.middlewareToPromise(mock(passedInArgument));
+            middlewareAsync().then(() => {
+                expect(mockDependency.call).toHaveBeenCalled();
+            });
+        });
+    });
```
Comment (fidian) [inline README.md:91]: Example seems a bit weird.  Not everyone uses factories, especially a middleware factory factory.

```
var middleware, middlewareAsync, req, res;

middleware = require("../lib/your-favorite-middleware");
middlewareAsync = jasmine.middlewareToPromise(middleware);

beforeEach(() => {
    req = makeFakeRequestObject();  // You write this
    res = makeFakeResponseObject();  // You write this
});
describe("a suite", () => {
    it("is easier with promises", () => {
        // If an argument is passed, the promise is rejected.
        // If the middleware throws, the promise is rejected.
        // The promise helper automatically handles both.
        // Just remember to `return` the promise.
        return middlewareAsync(req, res).then(() => {
            expect(res.send).toHaveBeenCalled();
        });
    });
    it("works in a traditional way", (done) => {
        // You must manually ensure the middleware does not throw.
        // Also you must manually ensure no argument was passed to
        // the `done` / `next` callback.
        // Don't forget to call `done()` when everything's complete.
        function doneWithMiddleware(arg) {
            expect(arg).not.toBeDefined();
            expect(res.send).toHaveBeenCalled();
            done();
        }
        jasmine.expect(() => {
            middleware(req, res, doneWithMiddleware);
        }).not.toThrow();
    });
});
```

File: README.md [hunk 23]
```diff
@@ -1 +1,116 @@
...
+
+Promises helper
+---------------
+
+This makes it much easier to use promises in node tests. Instead of having to remember to pass `done` in the `it` part of the testing in [Jasmine], you simple need to return the promise and the rest is handled like normal.
+
+```js
+// Before
+describe("a suite", () => {
+    it("a test", (done) => {
```
Comment (fidian) [inline README.md]: Code fences -> indentation

"a test" -> "does something"

File: README.md [hunk 24]
```diff
@@ -1 +1,116 @@
...
+// After
+describe("a suite", () => {
+    it("a test", () => {
+        return someClass.someMethodAsync().then((result) => {
+            expect(result).toBe("this");
+        });
+    });
+});
+```
+See how much simpler that is? You won't need to add `done` to the `it` or accidentally put `done` in the `describe` and wonder why your tests are failing when you've been writing tests all day.
```
Comment (fidian) [inline README.md]: Blank line to separate code from paragraph.

File: lib/promises.helper.js [hunk 25]
```diff
@@ -0,0 +1,54 @@
...
+                }
+            };
+        }
+
+        /* eslint no-invalid-this:"off" */
+        original.apply(this, args);
+    };
+
+    fn.patchedForPromises = true;
+    return fn;
```
Comment (fidian) [inline lib/promises.helper.js:43]: Newline above.

File: lib/fail.helper.js [hunk 26]
```diff
@@ -0,0 +1,28 @@
+"use strict";
+
+if (!jasmine.fail) {
+    /**
+     * Sets up the scenario to actually fail; `actual` and `expected`
+     * need to never match.
+     *
+     * @param {string} actual
+     * @param {string} expected
+     * @param {Function} used just for testing purposes
```
Comment (fidian) [inline lib/fail.helper.js]: The variable name `expectFn` is not on this line.  eslint should have barked loudly.

File: package.json [hunk 27]
```diff
@@ -0,0 +1,40 @@
+{
+    "name": "jasmine-node-helpers",
+    "version": "1.0.0",
+    "description": "Public and free software for helping test NodeJS projects using Jasmine-Node.",
+    "main": "/index.js",
+    "scripts": {
+        "test": "istanbul cover jasmine-node --captureExceptions spec/ && codecov .",
```
Comment (fidian) [inline package.json]: Add eslint to this script like opentoken.

File: package.json [hunk 28]
```diff
@@ -0,0 +1,40 @@
...
+        "test",
+        "helper",
+        "node"
+    ],
+    "author": "Tests Always Included Team",
+    "contributors": [
+        {
+            "name": "Christopher Tetreault",
+            "url": "https://github.com/AbsentSemicolon"
+        }
```
Comment (fidian) [inline package.json:29]: Mind if I am a contributor?  :-)

File: spec/middleware-to-promise.helper.spec.js [hunk 29]
```diff
@@ -0,0 +1,79 @@
...
+        var res, req, server;
+
+        beforeEach(() => {
+            server = require("./mock/server-mock")();
+        });
+        it("should pass and run call", () => {
+            var middlewareAsync;
+
+            middlewareAsync = jasmine.middlewareToPromise(factory("pass")(server));
+            return middlewareAsync().then(() => {
```
Comment (fidian) [inline spec/middleware-to-promise.helper.spec.js:57]: Newline above "return", here and other places.

File: spec/mock/expect-mock.js [hunk 30]
```diff
@@ -0,0 +1,17 @@
+"use strict";
+
+module.exports = (thing) => {
+
```
Comment (fidian) [inline spec/mock/expect-mock.js]: Whoa, one too many newlines.  I didn't expect to see this today.

File: spec/promises.helper.spec.js [hunk 31]
```diff
@@ -0,0 +1,40 @@
...
+            check = 0;
+
+            return new Promise((resolve) => {
+                setTimeout(() => {
+                    check = 1;
+                    resolve();
+                }, 100);
+            });
+        });
+        it("waited for the beforeeach to finish", () => {
```
Comment (fidian) [inline spec/promises.helper.spec.js]: Please capitalize the 5th "e" on this line.

File: spec/promises.helper.spec.js [hunk 32]
```diff
@@ -0,0 +1,40 @@
...
+            expect(check).toBe(1);
+            return new Promise((resolve) => {
```
Comment (fidian) [inline spec/promises.helper.spec.js:21]: Newline above `return`.

File: .eslint.yaml (hunk 1 above)
Comment (AbsentSemicolon) [reply .eslint.yaml]: Didn't want to take a straight copy of it at first, so I'll update this from the open token one.

File: .eslintignore (hunk 2 above)
Comment (AbsentSemicolon) [reply .eslintignore]: Not sure, was working on a few repos at the same time, so probably got confused, but I've updated it now.

File: lib/fail.helper.js (hunk 26 above)
Comment (AbsentSemicolon) [reply lib/fail.helper.js]: Strange it did not complain about this.

File: package.json (hunk 27 above)
Comment (AbsentSemicolon) [reply package.json]: This is added.

File: package.json (hunk 28 above)
Comment (AbsentSemicolon) [reply package.json:29]: I added you as a contributor.

File: lib/promises.helper.js (hunk 25 above)
Comment (AbsentSemicolon) [reply lib/promises.helper.js:43]: Linter doesn't seem to be working properly, but I got this changed.

File: CONTRIBUTING.md (hunk 4 above)
Comment (lnagle) [reply CONTRIBUTING.md]: May want to change check to checking.

File: CONTRIBUTING.md [hunk 33]
```diff
@@ -0,0 +1,23 @@
...
+1. [ESLint] is installed and you can utilize the [ESLint] configuration created in this repo. If you are unfamiliar with [ESLint], check out the site and see how great it is at keeping code consistent and check for some bad lines of code.
+2. [EditorConfig] is installed. Again, if unfamiliar with this tool, please take a few moments to familiarize yourself with it as this is also a handy tool.
+
+## Forking the Repository
+1. Fork the repository from https://github/tests-always-included/jasmine-node-helpers. Please do not fork from a fork.
+2. Create a branch off of `master`. You don't want to work on `master` as this can cause issues later on.
+3. Write the code as you normally would, but make sure to adhere to the guidelines set out in the [ESLint] and [EditorConfig] guides.
+    * You should also include JSDoc blocks where applicable. [ESLint] will warn you where they should be and what is missing from them.
+4. Write tests. We want to maintain a 100% code coverage, so your pull request probably won't get approved if something isn't tested. This includes `else` branches of `if` statements as well.
```
Comment (lnagle) [inline CONTRIBUTING.md:23]: Might be worth mentioning coverage reports and how to verify that tests in a pull request are actually providing full coverage. Could be something as simple as "If you're not familiar with Codecov and coverage reports, take some time to research them."

File: README.md (hunk 15 above)
Comment (lnagle) [reply README.md]: I think I get your meaning with the first sentence, but it was a little hard to parse at first. "This is most useful when testing that a promise is getting to the reject as expected, and when you'd like the test to fail if the promise is resolved rather than rejected." may be a little bit more clear (assuming I'm understanding it).

File: spec/middleware-to-promise.helper.spec.js [hunk 34]
```diff
@@ -0,0 +1,79 @@
...
+                return resolve();
+            });
+        });
+        mockCall.otherCall.andCallFake(() => {
+            return new Promise((resolve, reject) => {
+                return reject();
+            });
+        });
+        factory = (secondResult) => {
+            second = createSpyObj("afsdf", [
```
Comment (lnagle) [inline spec/middleware-to-promise.helper.spec.js]: Do you need jasmine. to precede createSpyObj?

Is there a reason this is "afsdf"? More generally, is there a naming convention for the createSpyObj?

Comment (AbsentSemicolon) [reply spec/middleware-to-promise.helper.spec.js]: There is, it should be the name of variable, but my linter wasn't working so this wasn't caught.

File: spec/middleware-to-promise.helper.spec.js (hunk 29 above)
Comment (AbsentSemicolon) [reply spec/middleware-to-promise.helper.spec.js:57]: Linter was named wrong. Doh!

File: spec/mock/expect-mock.js (hunk 30 above)
Comment (AbsentSemicolon) [reply spec/mock/expect-mock.js]: Got this changed, linter complained about missing JsDoc as well once I got that fixed.

File: spec/promises.helper.spec.js (hunk 31 above)
Comment (AbsentSemicolon) [reply spec/promises.helper.spec.js]: Capitalized.

File: spec/promises.helper.spec.js (hunk 32 above)
Comment (AbsentSemicolon) [reply spec/promises.helper.spec.js:21]: New line added.

File: README.md (hunk 24 above)
Comment (AbsentSemicolon) [reply README.md]: Added.

File: README.md (hunk 23 above)
Comment (AbsentSemicolon) [reply README.md]: Using spacing and changed the test name.

File: README.md (hunk 20 above)
Comment (AbsentSemicolon) [reply README.md]: Added indentation.

File: README.md (hunk 19 above)
Comment (AbsentSemicolon) [reply README.md]: Hopefully made this make more sense.

File: README.md (hunk 18 above)
Comment (AbsentSemicolon) [reply README.md]: Added.

File: README.md (hunk 21 above)
Comment (AbsentSemicolon) [reply README.md]: Changed this so it makes more sense when looking at example.

File: README.md (hunk 17 above)
Comment (AbsentSemicolon) [reply README.md]: Showing arguments.

File: README.md (hunk 16 above)
Comment (AbsentSemicolon) [reply README.md]: Using indentation

File: README.md (hunk 15 above)
Comment (AbsentSemicolon) [reply README.md]: Hopefully made the description make more sense.

File: README.md (hunk 14 above)
Comment (AbsentSemicolon) [reply README.md]: Made it this way.

File: README.md (hunk 12 above)
Comment (AbsentSemicolon) [reply README.md]: Made it use indentation.

File: README.md (hunk 13 above)
Comment (AbsentSemicolon) [reply README.md]: Added change directory.

File: README.md (hunk 11 above)
Comment (AbsentSemicolon) [reply README.md]: Yeah, made indentation. Need to get that style guide out in a separate repo.

File: README.md (hunk 10 above)
Comment (AbsentSemicolon) [reply README.md:21]: Added.

File: README.md (hunk 9 above)
Comment (AbsentSemicolon) [reply README.md]: Added and changed.

File: README.md (hunk 8 above)
Comment (AbsentSemicolon) [reply README.md:4]: Added.

File: CONTRIBUTING.md (hunk 7 above)
Comment (AbsentSemicolon) [reply CONTRIBUTING.md]: Made this a link.

File: CONTRIBUTING.md (hunk 33 above)
Comment (AbsentSemicolon) [reply CONTRIBUTING.md:23]: Added description for Codecov and links.

File: CONTRIBUTING.md (hunk 5 above)
Comment (AbsentSemicolon) [reply CONTRIBUTING.md]: Added.

File: CONTRIBUTING.md (hunk 4 above)
Comment (AbsentSemicolon) [reply CONTRIBUTING.md]: Changed wording.

File: CONTRIBUTING.md (hunk 3 above)
Comment (AbsentSemicolon) [reply CONTRIBUTING.md]: I used hyphens in the readme but not here.

File: CONTRIBUTING.md (hunk 6 above)
Comment (AbsentSemicolon) [reply CONTRIBUTING.md:8]: Added.

**Comment by lnagle** [comment]
Reviewed and looks great!

**Comment by fidian** [comment]
Nice.  Merging.
